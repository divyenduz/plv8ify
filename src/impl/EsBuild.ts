import { build } from 'esbuild'
import { injectable } from 'inversify'
import nodeExternals from 'webpack-node-externals'
import "reflect-metadata"

class BundlerError extends Error {}

// TODO: fixme, this is exported only for tests, is that needed?
@injectable()
export class EsBuild implements Bundler {
  async bundle({ inputFile, globalName }: BundleArgs) {
    const esbuildResult = await build({
      entryPoints: [inputFile],
      globalName,
      platform: 'browser',
      external: [nodeExternals()],
      bundle: true,
      write: false,
      target: 'es2015',
    }).catch(() => new BundlerError('esbuild failed'))

    if (esbuildResult instanceof Error) {
      throw esbuildResult
    }

    const esbuildFile = esbuildResult.outputFiles.find((_) => true)
    const bundlesJs = esbuildFile.text
    return bundlesJs
  }
}
