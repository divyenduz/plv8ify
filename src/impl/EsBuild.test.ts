import tap from 'tap'

import { EsBuild } from './EsBuild'

tap.test('getBundleJs - inline mode', async (t) => {
  const esbuild = new EsBuild()
  const js = await esbuild.bundle({
    inputFile: './src/test-fixtures/input.fixture.ts',
    globalName: 'plv8ify',
  })
  t.matchSnapshot(js)
})

tap.test('getBundleJs - bad syntax', async (t) => {
  const esbuild = new EsBuild()
  const js = esbuild.bundle({
    inputFile: './src/test-fixtures/bad.fixture.ts',
    globalName: 'plv8ify',
  })
  t.rejects(js)
})

tap.test('getBundleJs - newline template', async (t) => {
  const esbuild = new EsBuild()
  const js = await esbuild.bundle({
    inputFile: './src/test-fixtures/newline-template.fixture.ts',
    globalName: 'plv8ify',
  })
  t.matchSnapshot(js)
})

tap.test('getBundleJs - newline string', async (t) => {
  const esbuild = new EsBuild()
  const js = await esbuild.bundle({
    inputFile: './src/test-fixtures/newline-string.fixture.ts',
    globalName: 'plv8ify',
  })
  t.matchSnapshot(js)
})
