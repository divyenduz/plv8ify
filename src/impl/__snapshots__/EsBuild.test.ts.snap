// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`EsBuild tests getBundleJs - inline mode 1`] = `
"// src/test-fixtures/input.fixture.ts
function sayHello(toWhom) {
  const greet = \`hello \` + toWhom;
  return greet;
}
export {
  sayHello
};
"
`;

exports[`EsBuild tests getBundleJs - bad syntax 1`] = `[BundlerError: esbuild failed]`;

exports[`EsBuild tests getBundleJs - newline template 1`] = `
"// src/test-fixtures/newline-template.fixture.ts
function sayHello() {
  const greetTemplate = \`hello 
 world\`;
  return greetTemplate;
}
export {
  sayHello
};
"
`;

exports[`EsBuild tests getBundleJs - newline string 1`] = `
"// src/test-fixtures/newline-string.fixture.ts
function sayHello() {
  const greetString = "hello\\nworld";
  return greetString;
}
export {
  sayHello
};
"
`;
